File: src/token_log.ts
import { compareCodePoint, canonicalJson, stableHash } from "./canonical.js";
import { BlockToken, DeltaToken } from "./reversible.js";

export const CIRCULAR_TOKEN_LOG_CERTIFICATE_SCHEMA = "trwm.circular_token_log_certificate.v1";

export interface CircularTokenLogCertificate {
  schemaVersion: string;
  capacity: number;
  totalTokenCount: number;
  compactedPrefixCount: number;
  suffixCount: number;
  compactedDeltaCount: number;
  compactedTokens: DeltaToken[];
  suffixTokens: DeltaToken[];
  baseStateHash: string;
  compactedStateHash: string;
  finalStateHash: string;
  compactedTokenHash: string;
  suffixTokenHash: string;
  certificateHash: string;
}

export class CircularTokenLog {
  capacity: number;
  baseState: Record<string, unknown>;
  totalTokenCount: number;
  compactedPrefixCount: number;
  compactedTokens: DeltaToken[];
  suffixTokens: DeltaToken[];

  constructor(params: {
    capacity: number;
    baseState: Record<string, unknown>;
    totalTokenCount?: number;
    compactedPrefixCount?: number;
    compactedTokens?: DeltaToken[];
    suffixTokens?: DeltaToken[];
  }) {
    if (!positiveInteger(params.capacity)) {
      throw new RangeError("circular token log capacity must be a positive integer");
    }
    this.capacity = params.capacity;
    this.baseState = { ...params.baseState };
    this.totalTokenCount = params.totalTokenCount ?? 0;
    this.compactedPrefixCount = params.compactedPrefixCount ?? 0;
    this.compactedTokens = [...(params.compactedTokens ?? [])];
    this.suffixTokens = [...(params.suffixTokens ?? [])];
    if (this.suffixTokens.length > this.capacity) {
      throw new RangeError("suffix length exceeds circular token log capacity");
    }
  }

  static fromTokens(baseState: Record<string, unknown>, tokens: Iterable<DeltaToken>, capacity: number): CircularTokenLog {
    let log = new CircularTokenLog({ capacity, baseState });
    for (const token of tokens) {
      log = log.append(token);
    }
    return log;
  }

  append(token: DeltaToken): CircularTokenLog {
    let nextSuffix = [...this.suffixTokens, token];
    let compactedTokens = this.compactedTokens;
    let compactedPrefixCount = this.compactedPrefixCount;
    if (nextSuffix.length > this.capacity) {
      const evicted = nextSuffix[0];
      nextSuffix = nextSuffix.slice(1);
      compactedTokens = compactTokenPrefix(this.baseState, [...compactedTokens, evicted]);
      compactedPrefixCount += 1;
    }
    return new CircularTokenLog({
      capacity: this.capacity,
      baseState: this.baseState,
      totalTokenCount: this.totalTokenCount + 1,
      compactedPrefixCount,
      compactedTokens,
      suffixTokens: nextSuffix,
    });
  }

  replay(): Record<string, unknown> {
    return replayCircularTokenLog(this.baseState, this.compactedTokens, this.suffixTokens);
  }

  compactedState(): Record<string, unknown> {
    return this.compactedTokens.length > 0 ? BlockToken.of(this.compactedTokens).apply(this.baseState) : { ...this.baseState };
  }

  async certificate(): Promise<CircularTokenLogCertificate> {
    const compactedState = this.compactedState();
    const finalState = this.replay();
    const pending: CircularTokenLogCertificate = {
      schemaVersion: CIRCULAR_TOKEN_LOG_CERTIFICATE_SCHEMA,
      capacity: this.capacity,
      totalTokenCount: this.totalTokenCount,
      compactedPrefixCount: this.compactedPrefixCount,
      suffixCount: this.suffixTokens.length,
      compactedDeltaCount: this.compactedTokens.length,
      compactedTokens: this.compactedTokens,
      suffixTokens: this.suffixTokens,
      baseStateHash: await stableHash(this.baseState),
      compactedStateHash: await stableHash(compactedState),
      finalStateHash: await stableHash(finalState),
      compactedTokenHash: await tokenSequenceHash(this.compactedTokens),
      suffixTokenHash: await tokenSequenceHash(this.suffixTokens),
      certificateHash: "",
    };
    return { ...pending, certificateHash: await circularTokenLogCertificateHash(pending) };
  }
}

export function compactTokenPrefix(baseState: Record<string, unknown>, tokens: Iterable<DeltaToken>): DeltaToken[] {
  const base = { ...baseState };
  let current: Record<string, unknown> = { ...base };
  const touched = new Set<string>();
  for (const token of tokens) {
    current = token.apply(current);
    for (const key of token.readSet) touched.add(key);
    for (const key of token.writeSet) touched.add(key);
  }
  const compacted: DeltaToken[] = [];
  for (const key of Array.from(touched).sort(compareCodePoint)) {
    if (!Object.prototype.hasOwnProperty.call(base, key)) {
      throw new Error(`cannot compact missing base key: ${key}`);
    }
    if (canonicalJson(current[key]) !== canonicalJson(base[key])) {
      compacted.push(new DeltaToken(key, base[key], current[key]));
    }
  }
  return compacted;
}

export function replayCircularTokenLog(
  baseState: Record<string, unknown>,
  compactedTokens: Iterable<DeltaToken>,
  suffixTokens: Iterable<DeltaToken>,
): Record<string, unknown> {
  let current = { ...baseState };
  const compacted = Array.from(compactedTokens);
  const suffix = Array.from(suffixTokens);
  if (compacted.length > 0) {
    current = BlockToken.of(compacted).apply(current);
  }
  if (suffix.length > 0) {
    current = BlockToken.of(suffix).apply(current);
  }
  return current;
}

export async function buildCircularTokenLogCertificate(
  baseState: Record<string, unknown>,
  tokens: Iterable<DeltaToken>,
  options: { capacity: number },
): Promise<CircularTokenLogCertificate> {
  const rows = Array.from(tokens);
  const log = CircularTokenLog.fromTokens(baseState, rows, options.capacity);
  const fullState = rows.length > 0 ? BlockToken.of(rows).apply(baseState) : { ...baseState };
  const compactedState = log.replay();
  if (canonicalJson(fullState) !== canonicalJson(compactedState)) {
    throw new Error("circular token log compaction changed replay state");
  }
  return log.certificate();
}

export async function auditCircularTokenLog(
  baseState: Record<string, unknown>,
  certificate: CircularTokenLogCertificate,
  originalTokens?: Iterable<DeltaToken>,
): Promise<boolean> {
  if (!await validateCircularTokenLogCertificate(certificate)) {
    return false;
  }
  try {
    if (await stableHash(baseState) !== certificate.baseStateHash) {
      return false;
    }
    const compactedState = certificate.compactedTokens.length > 0
      ? BlockToken.of(certificate.compactedTokens).apply(baseState)
      : { ...baseState };
    const finalState = replayCircularTokenLog(baseState, certificate.compactedTokens, certificate.suffixTokens);
    const inverseState = inverseReplay(finalState, certificate.compactedTokens, certificate.suffixTokens);
    if (await stableHash(compactedState) !== certificate.compactedStateHash) {
      return false;
    }
    if (await stableHash(finalState) !== certificate.finalStateHash) {
      return false;
    }
    if (canonicalJson(inverseState) !== canonicalJson(baseState)) {
      return false;
    }
    if (originalTokens) {
      const rebuilt = await buildCircularTokenLogCertificate(baseState, originalTokens, { capacity: certificate.capacity });
      if (rebuilt.certificateHash !== certificate.certificateHash) {
        return false;
      }
    }
    return true;
  } catch (_error) {
    return false;
  }
}

export async function validateCircularTokenLogCertificate(certificate: CircularTokenLogCertificate): Promise<boolean> {
  try {
    if (certificate.schemaVersion !== CIRCULAR_TOKEN_LOG_CERTIFICATE_SCHEMA) {
      return false;
    }
    if (!positiveInteger(certificate.capacity)
      || !nonNegativeInteger(certificate.totalTokenCount)
      || !nonNegativeInteger(certificate.compactedPrefixCount)
      || !nonNegativeInteger(certificate.suffixCount)
      || !nonNegativeInteger(certificate.compactedDeltaCount)) {
      return false;
    }
    if (certificate.suffixCount !== certificate.suffixTokens.length) {
      return false;
    }
    if (certificate.compactedDeltaCount !== certificate.compactedTokens.length) {
      return false;
    }
    if (certificate.suffixCount > certificate.capacity) {
      return false;
    }
    if (certificate.totalTokenCount !== certificate.compactedPrefixCount + certificate.suffixCount) {
      return false;
    }
    if (certificate.compactedTokenHash !== await tokenSequenceHash(certificate.compactedTokens)) {
      return false;
    }
    if (certificate.suffixTokenHash !== await tokenSequenceHash(certificate.suffixTokens)) {
      return false;
    }
    return certificate.certificateHash === await circularTokenLogCertificateHash(certificate);
  } catch (_error) {
    return false;
  }
}

export async function circularTokenLogCertificateHash(
  certificate: CircularTokenLogCertificate | Record<string, unknown>,
): Promise<string> {
  const { certificateHash: _certificateHash, ...withoutHash } = certificate;
  return stableHash(withoutHash);
}

export async function tokenSequenceHash(tokens: Iterable<DeltaToken>): Promise<string> {
  return stableHash(Array.from(tokens));
}

export async function randomizedCircularTokenLogTrials(options: {
  seed?: number;
  trials?: number;
  keyCount?: number;
  tokenCount?: number;
  capacity?: number;
} = {}): Promise<{ trials: number; mismatches: number }> {
  const seed = options.seed ?? 17;
  const trials = options.trials ?? 64;
  const keyCount = options.keyCount ?? 5;
  const tokenCount = options.tokenCount ?? 14;
  const capacity = options.capacity ?? 4;
  if (!nonNegativeInteger(trials) || !positiveInteger(keyCount) || !nonNegativeInteger(tokenCount) || !positiveInteger(capacity)) {
    throw new RangeError("invalid randomized circular token log trial parameters");
  }
  const rng = mulberry32(seed >>> 0);
  let mismatches = 0;
  for (let trial = 0; trial < trials; trial += 1) {
    const state: Record<string, number> = {};
    for (let idx = 0; idx < keyCount; idx += 1) {
      state[`k${idx}`] = 0;
    }
    const current = { ...state };
    const tokens: DeltaToken[] = [];
    for (let idx = 0; idx < tokenCount; idx += 1) {
      const key = `k${Math.floor(rng() * keyCount)}`;
      const before = current[key];
      let after = before + Math.floor(rng() * 6) - 2;
      if (after === before) {
        after += 1;
      }
      tokens.push(new DeltaToken(key, before, after));
      current[key] = after;
    }
    const certificate = await buildCircularTokenLogCertificate(state, tokens, { capacity });
    if (!await auditCircularTokenLog(state, certificate, tokens)) {
      mismatches += 1;
    }
  }
  return { trials, mismatches };
}

function inverseReplay(
  finalState: Record<string, unknown>,
  compactedTokens: Iterable<DeltaToken>,
  suffixTokens: Iterable<DeltaToken>,
): Record<string, unknown> {
  let current = { ...finalState };
  const suffix = Array.from(suffixTokens);
  const compacted = Array.from(compactedTokens);
  if (suffix.length > 0) {
    current = BlockToken.of(suffix).inverse().apply(current);
  }
  if (compacted.length > 0) {
    current = BlockToken.of(compacted).inverse().apply(current);
  }
  return current;
}

function positiveInteger(value: number): boolean {
  return Number.isInteger(value) && value > 0;
}

function nonNegativeInteger(value: number): boolean {
  return Number.isInteger(value) && value >= 0;
}

function mulberry32(seed: number): () => number {
  let state = seed >>> 0;
  return () => {
    state = (state + 0x6D2B79F5) >>> 0;
    let value = state;
    value = Math.imul(value ^ (value >>> 15), value | 1);
    value ^= value + Math.imul(value ^ (value >>> 7), value | 61);
    return ((value ^ (value >>> 14)) >>> 0) / 4294967296;
  };
}

File: trwm/token_log.py
from __future__ import annotations

from dataclasses import asdict, dataclass
import random
from typing import Any, Iterable, Mapping

from .core import stable_hash
from .reversible import BlockToken, DeltaToken


CIRCULAR_TOKEN_LOG_CERTIFICATE_SCHEMA = "trwm.circular_token_log_certificate.v1"


@dataclass(frozen=True)
class CircularTokenLogCertificate:
    schema_version: str
    capacity: int
    total_token_count: int
    compacted_prefix_count: int
    suffix_count: int
    compacted_delta_count: int
    compacted_tokens: tuple[DeltaToken, ...]
    suffix_tokens: tuple[DeltaToken, ...]
    base_state_hash: str
    compacted_state_hash: str
    final_state_hash: str
    compacted_token_hash: str
    suffix_token_hash: str
    certificate_hash: str = ""

    def __post_init__(self) -> None:
        if self.schema_version != CIRCULAR_TOKEN_LOG_CERTIFICATE_SCHEMA:
            raise ValueError(f"invalid circular token log schema: {self.schema_version}")
        if any(
            not isinstance(value, int) or isinstance(value, bool) or value < 0
            for value in (
                self.capacity,
                self.total_token_count,
                self.compacted_prefix_count,
                self.suffix_count,
                self.compacted_delta_count,
            )
        ):
            raise ValueError("circular token log counts must be non-negative integers")
        if self.capacity <= 0:
            raise ValueError("circular token log capacity must be positive")
        object.__setattr__(self, "compacted_tokens", tuple(self.compacted_tokens))
        object.__setattr__(self, "suffix_tokens", tuple(self.suffix_tokens))
        if not self.certificate_hash:
            object.__setattr__(self, "certificate_hash", circular_token_log_certificate_hash(self))

    def without_hash(self) -> dict[str, Any]:
        data = asdict(self)
        data.pop("certificate_hash", None)
        return data


@dataclass(frozen=True)
class CircularTokenLog:
    capacity: int
    base_state: Mapping[str, Any]
    total_token_count: int = 0
    compacted_prefix_count: int = 0
    compacted_tokens: tuple[DeltaToken, ...] = ()
    suffix_tokens: tuple[DeltaToken, ...] = ()

    def __post_init__(self) -> None:
        if not isinstance(self.capacity, int) or isinstance(self.capacity, bool) or self.capacity <= 0:
            raise ValueError("circular token log capacity must be a positive integer")
        object.__setattr__(self, "base_state", dict(self.base_state))
        object.__setattr__(self, "compacted_tokens", tuple(self.compacted_tokens))
        object.__setattr__(self, "suffix_tokens", tuple(self.suffix_tokens))
        if len(self.suffix_tokens) > self.capacity:
            raise ValueError("suffix length exceeds circular token log capacity")

    @classmethod
    def from_tokens(cls, base_state: Mapping[str, Any], tokens: Iterable[DeltaToken], capacity: int) -> "CircularTokenLog":
        log = cls(capacity=capacity, base_state=base_state)
        for token in tokens:
            log = log.append(token)
        return log

    def append(self, token: DeltaToken) -> "CircularTokenLog":
        next_suffix = self.suffix_tokens + (token,)
        compacted_tokens = self.compacted_tokens
        compacted_prefix_count = self.compacted_prefix_count
        if len(next_suffix) > self.capacity:
            evicted = next_suffix[0]
            next_suffix = next_suffix[1:]
            compacted_tokens = compact_token_prefix(self.base_state, compacted_tokens + (evicted,))
            compacted_prefix_count += 1
        return CircularTokenLog(
            capacity=self.capacity,
            base_state=self.base_state,
            total_token_count=self.total_token_count + 1,
            compacted_prefix_count=compacted_prefix_count,
            compacted_tokens=compacted_tokens,
            suffix_tokens=next_suffix,
        )

    def replay(self) -> dict[str, Any]:
        return replay_circular_token_log(self.base_state, self.compacted_tokens, self.suffix_tokens)

    def compacted_state(self) -> dict[str, Any]:
        return BlockToken.of(self.compacted_tokens).apply(self.base_state) if self.compacted_tokens else dict(self.base_state)

    def certificate(self) -> CircularTokenLogCertificate:
        compacted_state = self.compacted_state()
        final_state = replay_circular_token_log(self.base_state, self.compacted_tokens, self.suffix_tokens)
        return CircularTokenLogCertificate(
            schema_version=CIRCULAR_TOKEN_LOG_CERTIFICATE_SCHEMA,
            capacity=self.capacity,
            total_token_count=self.total_token_count,
            compacted_prefix_count=self.compacted_prefix_count,
            suffix_count=len(self.suffix_tokens),
            compacted_delta_count=len(self.compacted_tokens),
            compacted_tokens=self.compacted_tokens,
            suffix_tokens=self.suffix_tokens,
            base_state_hash=stable_hash(dict(self.base_state)),
            compacted_state_hash=stable_hash(compacted_state),
            final_state_hash=stable_hash(final_state),
            compacted_token_hash=token_sequence_hash(self.compacted_tokens),
            suffix_token_hash=token_sequence_hash(self.suffix_tokens),
        )


def compact_token_prefix(base_state: Mapping[str, Any], tokens: Iterable[DeltaToken]) -> tuple[DeltaToken, ...]:
    base = dict(base_state)
    current = dict(base)
    touched: set[str] = set()
    for token in tokens:
        current = token.apply(current)
        touched.update(token.read_set)
        touched.update(token.write_set)
    compacted: list[DeltaToken] = []
    for key in sorted(touched):
        if key not in base:
            raise ValueError(f"cannot compact missing base key: {key}")
        if current[key] != base[key]:
            compacted.append(DeltaToken(key, base[key], current[key]))
    return tuple(compacted)


def replay_circular_token_log(
    base_state: Mapping[str, Any],
    compacted_tokens: Iterable[DeltaToken],
    suffix_tokens: Iterable[DeltaToken],
) -> dict[str, Any]:
    current = dict(base_state)
    for block in (tuple(compacted_tokens), tuple(suffix_tokens)):
        if block:
            current = BlockToken.of(block).apply(current)
    return current


def build_circular_token_log_certificate(
    base_state: Mapping[str, Any],
    tokens: Iterable[DeltaToken],
    *,
    capacity: int,
) -> CircularTokenLogCertificate:
    rows = tuple(tokens)
    log = CircularTokenLog.from_tokens(base_state, rows, capacity)
    full_state = BlockToken.of(rows).apply(base_state) if rows else dict(base_state)
    compacted_state = log.replay()
    if full_state != compacted_state:
        raise AssertionError("circular token log compaction changed replay state")
    return log.certificate()


def audit_circular_token_log(
    base_state: Mapping[str, Any],
    certificate: CircularTokenLogCertificate,
    original_tokens: Iterable[DeltaToken] | None = None,
) -> bool:
    if not validate_circular_token_log_certificate(certificate):
        return False
    try:
        if stable_hash(dict(base_state)) != certificate.base_state_hash:
            return False
        compacted_state = BlockToken.of(certificate.compacted_tokens).apply(base_state) if certificate.compacted_tokens else dict(base_state)
        final_state = replay_circular_token_log(base_state, certificate.compacted_tokens, certificate.suffix_tokens)
        inverse_state = _inverse_replay(final_state, certificate.compacted_tokens, certificate.suffix_tokens)
        if stable_hash(compacted_state) != certificate.compacted_state_hash:
            return False
        if stable_hash(final_state) != certificate.final_state_hash:
            return False
        if inverse_state != dict(base_state):
            return False
        if original_tokens is not None:
            rebuilt = build_circular_token_log_certificate(base_state, original_tokens, capacity=certificate.capacity)
            if rebuilt.certificate_hash != certificate.certificate_hash:
                return False
        return True
    except Exception:
        return False


def validate_circular_token_log_certificate(certificate: CircularTokenLogCertificate) -> bool:
    try:
        if certificate.schema_version != CIRCULAR_TOKEN_LOG_CERTIFICATE_SCHEMA:
            return False
        if certificate.suffix_count != len(certificate.suffix_tokens):
            return False
        if certificate.compacted_delta_count != len(certificate.compacted_tokens):
            return False
        if certificate.suffix_count > certificate.capacity:
            return False
        if certificate.total_token_count != certificate.compacted_prefix_count + certificate.suffix_count:
            return False
        if certificate.compacted_token_hash != token_sequence_hash(certificate.compacted_tokens):
            return False
        if certificate.suffix_token_hash != token_sequence_hash(certificate.suffix_tokens):
            return False
        return certificate.certificate_hash == circular_token_log_certificate_hash(certificate)
    except Exception:
        return False


def circular_token_log_certificate_hash(certificate: CircularTokenLogCertificate | Mapping[str, Any]) -> str:
    if isinstance(certificate, CircularTokenLogCertificate):
        data = certificate.without_hash()
    else:
        data = dict(certificate)
        data.pop("certificate_hash", None)
    return stable_hash(data)


def token_sequence_hash(tokens: Iterable[DeltaToken]) -> str:
    return stable_hash(tuple(tokens))


def randomized_circular_token_log_trials(
    *,
    seed: int = 17,
    trials: int = 64,
    key_count: int = 5,
    token_count: int = 14,
    capacity: int = 4,
) -> tuple[int, int]:
    if trials < 0 or key_count <= 0 or token_count < 0 or capacity <= 0:
        raise ValueError("invalid randomized circular token log trial parameters")
    rng = random.Random(seed)
    mismatches = 0
    for _trial in range(trials):
        state = {f"k{idx}": 0 for idx in range(key_count)}
        current = dict(state)
        tokens: list[DeltaToken] = []
        for _idx in range(token_count):
            key = f"k{rng.randrange(key_count)}"
            before = current[key]
            after = before + rng.randint(-2, 3)
            if after == before:
                after += 1
            tokens.append(DeltaToken(key, before, after))
            current[key] = after
        certificate = build_circular_token_log_certificate(state, tokens, capacity=capacity)
        if not audit_circular_token_log(state, certificate, tokens):
            mismatches += 1
    return trials, mismatches


def _inverse_replay(
    final_state: Mapping[str, Any],
    compacted_tokens: Iterable[DeltaToken],
    suffix_tokens: Iterable[DeltaToken],
) -> dict[str, Any]:
    current = dict(final_state)
    suffix = tuple(suffix_tokens)
    compacted = tuple(compacted_tokens)
    if suffix:
        current = BlockToken.of(suffix).inverse().apply(current)
    if compacted:
        current = BlockToken.of(compacted).inverse().apply(current)
    return current
